### PR TITLE
DOC: Update the DataFrame.join() docstring 

### PR DESCRIPTION
**Pull request by kantologist**
Checklist for the pandas documentation sprint (ignore this if you are doing
an unrelated PR):

- [x] PR title is "DOC: update the <your-function-or-method> docstring"
- [x] The validation script passes: `scripts/validate_docstrings.py <your-function-or-method>`
- [x] The PEP8 style check passes: `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] The html version looks good: `python doc/make.py --single <your-function-or-method>`
- [x] It has been proofread on language by another sprint participant

Please include the output of the validation script below between the "```" ticks:

```
################################################################################
###################### Docstring (pandas.DataFrame.join)  ######################
################################################################################

Join columns with other DataFrame either on index or a key
column.

Efficiently Join multiple DataFrame objects by index at once by
passing a list.

Parameters
----------
other : DataFrame, Series with name field set, or list of DataFrame
    Index should be similar to one of the columns in this one. If a
    Series is passed, its name attribute must be set, and that will be
    used as the column name in the resulting joined DataFrame.
on : name, tuple/list of names, or array-like
    Column or index level name(s) in the caller to join on the index
    in `other`, otherwise joins index-on-index. If multiple
    values given, the `other` DataFrame must have a MultiIndex. Can
    pass an array as the join key if it is not already contained in
    the calling DataFrame. Like an Excel VLOOKUP operation.
how : {'left', 'right', 'outer', 'inner'}, default: 'left'
    How to handle the operation of the two objects.

    * left: use calling frame's index (or column if on is specified).
    * right: use other frame's index.
    * outer: form union of calling frame's index (or column if on is
      specified) with other frame's index, and sort it
      lexicographically.
    * inner: form intersection of calling frame's index (or column if
      on is specified) with other frame's index, preserving the order
      of the calling's one.
lsuffix : string
    Suffix to use from left frame's overlapping columns.
rsuffix : string
    Suffix to use from right frame's overlapping columns.
sort : boolean, default False
    Order result DataFrame lexicographically by the join key. If False,
    the order of the join key depends on the join type (how keyword).

Notes
-----
The on, lsuffix, and rsuffix options are not supported when passing
a list of DataFrame objects.

Support for specifying index levels as the `on` parameter was added
in version 0.23.0.

Examples
--------
>>> caller = pd.DataFrame({'A': ['A0', 'A1', 'A2', 'A3', 'A4', 'A5'],
...                      'key': ['K0', 'K1', 'K2', 'K3', 'K4', 'K5']})

>>> caller
    A key
0  A0  K0
1  A1  K1
2  A2  K2
3  A3  K3
4  A4  K4
5  A5  K5

>>> other = pd.DataFrame({'B': ['B0', 'B1', 'B2'],
...                       'key': ['K0', 'K1', 'K2']})

>>> other
    B key
0  B0  K0
1  B1  K1
2  B2  K2

Join DataFrames using their indexes.

>>> caller.join(other, lsuffix='_caller', rsuffix='_other')
    A key_caller    B key_other
0  A0         K0   B0        K0
1  A1         K1   B1        K1
2  A2         K2   B2        K2
3  A3         K3  NaN       NaN
4  A4         K4  NaN       NaN
5  A5         K5  NaN       NaN

If we want to join using the key columns, we need to set key to be
the index in both caller and other. The joined DataFrame will have
key as its index.

>>> caller.set_index('key').join(other.set_index('key'))
  A    B
key
K0   A0   B0
K1   A1   B1
K2   A2   B2
K3   A3  NaN
K4   A4  NaN
K5   A5  NaN

Another option to join using the key columns is to use the on
parameter. DataFrame.join always uses other's index but we can use any
column in the caller. This method preserves the original caller's
index in the result.

>>> caller.join(other.set_index('key'), on='key')
    A key    B
0  A0  K0   B0
1  A1  K1   B1
2  A2  K2   B2
3  A3  K3  NaN
4  A4  K4  NaN
5  A5  K5  NaN

See also
--------
DataFrame.merge : For column(s)-on-columns(s) operations

Returns
-------
joined : DataFrame

################################################################################
################################## Validation ##################################
################################################################################

Docstring for "pandas.DataFrame.join" correct. :)
```

If the validation script still gives errors, but you think there is a good reason
to deviate in this case (and there are certainly such cases), please state this
explicitly.

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -5177,53 +5177,55 @@ def append(self, other, ignore_index=False, verify_integrity=False):
...
         passing a list.
 
         Parameters
         ----------
         other : DataFrame, Series with name field set, or list of DataFrame
             Index should be similar to one of the columns in this one. If a
             Series is passed, its name attribute must be set, and that will be
-            used as the column name in the resulting joined DataFrame
+            used as the column name in the resulting joined DataFrame.
         on : name, tuple/list of names, or array-like
```
Comment (WillAyd) [inline pandas/core/frame.py]: The types could use a clean up here. Something like:

`str, list of str, tuple of (str,), or array-like`

File: pandas/core/frame.py [hunk 2]
```diff
@@ -5177,53 +5177,55 @@ def append(self, other, ignore_index=False, verify_integrity=False):
...
-        column. Efficiently Join multiple DataFrame objects by index at once by
+        Join columns with other DataFrame either on index or a key
+        column.
+
+        Efficiently Join multiple DataFrame objects by index at once by
         passing a list.
 
         Parameters
         ----------
         other : DataFrame, Series with name field set, or list of DataFrame
```
Comment (WillAyd) [inline pandas/core/frame.py]: Without trying to do to much, perhaps this line should just be `DataFrame, Series or list of DataFrame`. The detail around the name being set is in the comment below, and could maybe be reinforced in the `Notes` section

File: pandas/core/frame.py [hunk 3]
```diff
@@ -5177,53 +5177,55 @@ def append(self, other, ignore_index=False, verify_integrity=False):
...
             Index should be similar to one of the columns in this one. If a
             Series is passed, its name attribute must be set, and that will be
-            used as the column name in the resulting joined DataFrame
+            used as the column name in the resulting joined DataFrame.
         on : name, tuple/list of names, or array-like
             Column or index level name(s) in the caller to join on the index
```
Comment (WillAyd) [inline pandas/core/frame.py]: Can you think of any better ways to word this?

File: pandas/core/frame.py [hunk 4]
```diff
@@ -5177,53 +5177,55 @@ def append(self, other, ignore_index=False, verify_integrity=False):
...
             values given, the `other` DataFrame must have a MultiIndex. Can
             pass an array as the join key if it is not already contained in
-            the calling DataFrame. Like an Excel VLOOKUP operation
+            the calling DataFrame. Like an Excel VLOOKUP operation.
         how : {'left', 'right', 'outer', 'inner'}, default: 'left'
             How to handle the operation of the two objects.
 
-            * left: use calling frame's index (or column if on is specified)
-            * right: use other frame's index
+            * left: use calling frame's index (or column if on is specified).
```
Comment (WillAyd) [inline pandas/core/frame.py]: Understood why you did this, but generally you don't put periods at the end of bullet points

File: pandas/core/frame.py [hunk 5]
```diff
@@ -5246,46 +5248,41 @@ def join(self, other, on=None, how='left', lsuffix='', rsuffix='',
...
-
+            A key    B
+        0  A0  K0   B0
+        1  A1  K1   B1
+        2  A2  K2   B2
+        3  A3  K3  NaN
+        4  A4  K4  NaN
+        5  A5  K5  NaN
 
         See also
```
Comment (WillAyd) [inline pandas/core/frame.py]: Capitalize "Also"

**Review by WillAyd — CHANGES_REQUESTED** [review]
Noted a few improvements in various comments

File: pandas/core/frame.py [hunk 6]
```diff
@@ -5177,53 +5177,55 @@ def append(self, other, ignore_index=False, verify_integrity=False):
     def join(self, other, on=None, how='left', lsuffix='', rsuffix='',
              sort=False):
         """
-        Join columns with other DataFrame either on index or on a key
-        column. Efficiently Join multiple DataFrame objects by index at once by
+        Join columns with other DataFrame either on index or a key
```
Comment (jreback) [inline pandas/core/frame.py]: on the index or a key column. might be too long though.

File: pandas/core/frame.py [hunk 7]
```diff
@@ -5177,53 +5177,55 @@ def append(self, other, ignore_index=False, verify_integrity=False):
...
+        column.
+
+        Efficiently Join multiple DataFrame objects by index at once by
         passing a list.
```
Comment (jreback) [inline pandas/core/frame.py]: passing a list of DataFrames

File: pandas/core/frame.py [hunk 8]
```diff
@@ -5177,53 +5177,55 @@ def append(self, other, ignore_index=False, verify_integrity=False):
...
         Support for specifying index levels as the `on` parameter was added
-        in version 0.23.0
+        in version 0.23.0.
 
         Examples
         --------
-        >>> caller = pd.DataFrame({'key': ['K0', 'K1', 'K2', 'K3', 'K4', 'K5'],
-        ...                        'A': ['A0', 'A1', 'A2', 'A3', 'A4', 'A5']})
+        >>> caller = pd.DataFrame({'A': ['A0', 'A1', 'A2', 'A3', 'A4', 'A5'],
+        ...                      'key': ['K0', 'K1', 'K2', 'K3', 'K4', 'K5']})
```
Comment (jreback) [inline pandas/core/frame.py]: the original formatting was ok here

File: pandas/core/frame.py [hunk 9]
```diff
@@ -5246,46 +5248,41 @@ def join(self, other, on=None, how='left', lsuffix='', rsuffix='',
...
-
-        >>>      A    B
-            key
-            K0   A0   B0
-            K1   A1   B1
-            K2   A2   B2
-            K3   A3  NaN
-            K4   A4  NaN
-            K5   A5  NaN
+          A    B
```
Comment (jreback) [inline pandas/core/frame.py]: the column headers should be lined up

File: pandas/core/frame.py [hunk 10]
```diff
@@ -5246,46 +5248,41 @@ def join(self, other, on=None, how='left', lsuffix='', rsuffix='',
...
-
-        >>>     A key    B
-            0  A0  K0   B0
-            1  A1  K1   B1
-            2  A2  K2   B2
-            3  A3  K3  NaN
-            4  A4  K4  NaN
-            5  A5  K5  NaN
-
+            A key    B
```
Comment (jreback) [inline pandas/core/frame.py]: is this how it renders? I *think* there should be another space before the entire column of key

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py [hunk 11]
```diff
@@ -5295,22 +5295,24 @@ def append(self, other, ignore_index=False, verify_integrity=False):
     def join(self, other, on=None, how='left', lsuffix='', rsuffix='',
              sort=False):
         """
-        Join columns with other DataFrame either on index or on a key
-        column. Efficiently Join multiple DataFrame objects by index at once by
-        passing a list.
+        Join columns with other DataFrame either on a key column.
```
Comment (jreback) [inline pandas/core/frame.py]: Join with other Dataframe on the index or a key column.

File: pandas/core/frame.py [hunk 12]
```diff
@@ -5295,22 +5295,24 @@ def append(self, other, ignore_index=False, verify_integrity=False):
...
+
+        Efficiently Join multiple DataFrame objects by index at once by
```
Comment (jreback) [inline pandas/core/frame.py]: should be in the Notes section

File: pandas/core/frame.py [hunk 13]
```diff
@@ -5295,22 +5295,24 @@ def append(self, other, ignore_index=False, verify_integrity=False):
...
+        passing a list of DataFrames.
 
         Parameters
         ----------
-        other : DataFrame, Series with name field set, or list of DataFrame
+        other : DataFrame, Series, or list of DataFrame
```
Comment (jreback) [inline pandas/core/frame.py]: DataFrames

File: pandas/core/frame.py [hunk 14]
```diff
@@ -5295,22 +5295,24 @@ def append(self, other, ignore_index=False, verify_integrity=False):
...
             Index should be similar to one of the columns in this one. If a
-            Series is passed, its name attribute must be set, and that will be
-            used as the column name in the resulting joined DataFrame
+            Series is passed, its name attribute must be set, since
+            it would be used as the name of the appended
```
Comment (jreback) [inline pandas/core/frame.py]: appended -> added

File: pandas/core/frame.py [hunk 15]
```diff
@@ -5321,91 +5323,88 @@ def join(self, other, on=None, how='left', lsuffix='', rsuffix='',
...
+            the order of the join key depends on the join type (how keyword).
 
         Notes
         -----
-        on, lsuffix, and rsuffix options are not supported when passing a list
-        of DataFrame objects
+        The on, lsuffix, and rsuffix options are not supported when passing
+        a list of DataFrame objects.
+
+        If a Series is used as other parameter it must be with name field set.
```
Comment (jreback) [inline pandas/core/frame.py]: is must have the name attribute set.

File: pandas/core/frame.py [hunk 16]
```diff
@@ -5321,91 +5323,88 @@ def join(self, other, on=None, how='left', lsuffix='', rsuffix='',
...
 
         Support for specifying index levels as the `on` parameter was added
-        in version 0.23.0
+        in version 0.23.0.
 
         Examples
         --------
         >>> caller = pd.DataFrame({'key': ['K0', 'K1', 'K2', 'K3', 'K4', 'K5'],
-        ...                        'A': ['A0', 'A1', 'A2', 'A3', 'A4', 'A5']})
```
Comment (jreback) [inline pandas/core/frame.py]: was ok before

File: pandas/core/frame.py [hunk 17]
```diff
@@ -5321,91 +5323,88 @@ def join(self, other, on=None, how='left', lsuffix='', rsuffix='',
...
 
         >>> caller
-            A key
-        0  A0  K0
-        1  A1  K1
-        2  A2  K2
-        3  A3  K3
-        4  A4  K4
-        5  A5  K5
+          key   A
```
Comment (jreback) [inline pandas/core/frame.py]: was ok before

File: pandas/core/frame.py [hunk 18]
```diff
@@ -5321,91 +5323,88 @@ def join(self, other, on=None, how='left', lsuffix='', rsuffix='',
...
+        5  K5  A5
 
         >>> other = pd.DataFrame({'key': ['K0', 'K1', 'K2'],
         ...                       'B': ['B0', 'B1', 'B2']})
 
         >>> other
-            B key
-        0  B0  K0
-        1  B1  K1
-        2  B2  K2
```
Comment (jreback) [inline pandas/core/frame.py]: same

File: pandas/core/frame.py [hunk 19]
```diff
@@ -5321,91 +5323,88 @@ def join(self, other, on=None, how='left', lsuffix='', rsuffix='',
...
+        1  K1  A1   B1
+        2  K2  A2   B2
+        3  K3  A3  NaN
+        4  K4  A4  NaN
+        5  K5  A5  NaN
 
-        >>>     A key    B
-            0  A0  K0   B0
-            1  A1  K1   B1
-            2  A2  K2   B2
```
Comment (jreback) [inline pandas/core/frame.py]: leave all of these examples alone, they are fine

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by kantologist** [comment]
I have made the changes. Waiting for reviews.

**Comment by datapythonista** [comment]
@kantologist can you please restore the original examples, as per previous review?

**Comment by datapythonista** [comment]
Closing as stale.

**Pull request closed without merging**